### ISSUE TITLE
Bug in fixer for xUnit2014 when handling nested methods

### ISSUE THREAD

**Issue by carlossanlop** [xunit/xunit#2870]
I encountered this unit test code in an old project using a nested method to assert a Task:

```cs
using System;
using System.Threading.Tasks;
using Xunit;

public class MyTestClass
{
    private ValueTask GetValueTaskAsync()
    {
        return ValueTask.CompletedTask;
    }

    [Fact]
    public void MyTestMethod()
    {
        ValueTask vt = GetValueTaskAsync();
        MyNestedMethod(vt);

        void MyNestedMethod(ValueTask vt)
        {
            Assert.Throws<ArgumentOutOfRangeException>(() => vt.AsTask());
        }
    }
}
```

I got these two failures:

- `xUnit2014: Do not use Assert.Throws() to check for asynchronously thrown exceptions. Use Assert.ThrowsAsync instead.`
- `CS0619: Assert.Throws<T>(Func<Task>) is obsolete: You must call Assert.ThrowsAsync<T> (and await the result) when testing async code.`

### Actual

The code fix was suggested by xUnit2014. Unfortunately, it is wrong: it adds the missing method modifiers to the top method:

```cs
    [Fact]
    public async Task MyTestMethod()
    {
        ValueTask vt = GetValueTaskAsync();
        MyNestedMethod(vt);

        void MyNestedMethod(ValueTask vt)
        {
            await Assert.ThrowsAsync<ArgumentOutOfRangeException>(() => vt.AsTask());
        }
    }
```

### Expected

The correct fix should have added the modifiers to the nested method, and then should've added the await keyword to the nested method invocation:

```cs
    [Fact]
    public async Task MyTestMethod()
    {
        ValueTask vt = GetValueTaskAsync();
        await MyNestedMethod(vt);

        async Task MyNestedMethod(ValueTask vt)
        {
            await Assert.ThrowsAsync<ArgumentOutOfRangeException>(() => vt.AsTask());
        }
    }
```

**Comment by bradwilson** [comment]
The two failures are expected (you're calling a method marked with `Obsolete` so that the compiler will catch misuse rather than relying on the analyzer, but the analyzer is still required to be able to enable the fixer).

The incorrect method being modified is definitely a bug.

**Comment by bradwilson** [comment]
> Should fixing a deeply nested Assert.Throws invocation cascade upwards, adding async/await to all of the parent functions and invocations within the method (stopping only when a function has no invocations or when the outermost method level is reached)? Should we fully apply this to any recursive invocations that might be present?

You give two examples, but I think there's an in-between third option that actually provides a lot of value without a lot of extra work.

What if you add the `async` to `innerDelegate` and `await` the assertion (which is the original problem that xUnit2014 highlighted), and then just do the `async` piece all the way up (without adding any `await`s)? Then the compiler will help them out indicating where all the remaining issues are.

The fixed code would look something like this:

```csharp
[Fact]
public async Task PartlyFixedTestMethod()
{
    async Task<int> OuterLocalFunction()
    {
        Func<Task<bool>> outerDelegate = async () =>
        {
            async Task<string> InnerLocalFunction()
            {
                Func<Task> innerDelegate = async () =>
                {
                    await Assert.ThrowsAsync<Exception>(() => Task.Delay(0));
                };

                innerDelegate(); // Task is discarded and not awaited
                innerDelegate(); // Task is discarded and not awaited
                string message = InnerLocalFunction(); // Recursion
                return string.Empty;
            }

            string message = InnerLocalFunction();
            InnerLocalFunction();
            return false;
        };

        bool condition = outerDelegate();
        outerDelegate();
        return 0;
    }

    int number = OuterLocalFunction();
    OuterLocalFunction();
}
```

And now here's what I see inside VS code:

![image](https://github.com/xunit/xunit/assets/16944/5421f65a-9427-407c-9309-741454e6c5a4)

The yellow lines are CS1998 (`This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.`) and CS 4014 (`Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call.`), while the red lines are CS0029 (type incompatibilities like calling `InnerLocalFunction` and expecting to get back a `string`, but getting back a `Task<string>` instead because of the lack of `await`). This leads the developer to be able to fix the rest by themselves, with assistance from the compiler/IDE.

I think trying to do the _rest_ of the fixing for them is over-ambitious and subject to a lot of edge cases. We can just let them do it themselves; we found the fundamental problem, and they can fix the rest.

> Should async/await be added at all levels? Or, as @carlossanlop suggested as the second alternative expected fix, should async/await not be added to the inner levels, with Tasks being returned directly, and with async/await only added at the outermost method level?

Microsoft's advice is "always await, never just return", and while we agree to disagree (I think there are valid times to optimize), I think the decision to optimize is one to be left up to the developer. I'd say: always `await`.

**Comment by jared-chevalier** [comment]
> What if you add the `async` to `innerDelegate` and `await` the assertion (which is the original problem that xUnit2014 highlighted), and then just do the `async` piece all the way up (without adding any `await`s)? Then the compiler will help them out indicating where all the remaining issues are.

Yes, this makes sense and would certainly simplify the fixer compared to the other approach.

> Microsoft's advice is "always await, never just return", and while we agree to disagree (I think there are valid times to optimize), I think the decision to optimize is one to be left up to the developer. I'd say: always `await`.

Yes, I agree that it would be better to use `await` here and allow the developer to decide on potential optimizations.

Thanks for your help. I will proceed to work on implementing this as you have described.
